### ISSUE TITLE
Race in `Downward API volume should update annotations / labels on modification`

### ISSUE THREAD

**Issue by yifan-gu** [kubernetes/kubernetes#29633]
Ref https://github.com/kubernetes/kubernetes/issues/28462
Ref https://github.com/kubernetes/kubernetes/issues/28416

I believe these two flakes have the same root cause:

When there's a pod UPDATE event to kubelet, kubelet will invoke the `volume manager` to update the pod state to introduce the updated annotation:
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/volumemanager/volume_manager.go#L298

The `volume manager`:
- 1, Tells the `desiredStateOfWorld populator` to reprocess the volume mount in the next loop.
- 2, Mark the volume in `actualStateOfWorld` to be remounted.
  https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/volumemanager/volume_manager.go#L311-L312

However in the reconciler's reconciling loop, it does not always pick up the latest pod state with the latest annotation in the memory cache as the `populator` is running asynchronously.
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/volumemanager/reconciler/reconciler.go#L150

But when it reads the `actualStateOfWorld`, it does get notified that the volume needs to be remount, 
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/volumemanager/reconciler/reconciler.go#L223

So when remounting the volume, it ends up thinking that the annotation in the pod is not changed, and thus not modifying the content on the disk.

Although the annotation updates on the disk will happen eventually after another periodic pod SYNC event received by kubelet, but it will results in the timeout of the e2e tests.

cc @kubernetes/sig-storage

**Comment by yifan-gu** [comment]
This can be easily reproduced locally with such pod:
```yaml
apiVersion: v1
kind: Pod
metadata:
  annotations:
    builder: foo
  name: annotationannotationupdate
spec:
  containers:
  - command:
    - /mt
    - --break_on_expected_content=false
    - --retry_time=12000
    - --file_content_in_loop=/etc/annotations
    image: gcr.io/google_containers/mounttest:0.6
    image: busybox
    imagePullPolicy: IfNotPresent
    name: client-container
    resources: {}
    terminationMessagePath: /dev/termination-log
    volumeMounts:
    - mountPath: /opt
      name: podinfo
  dnsPolicy: ClusterFirst
  restartPolicy: Never
  securityContext: {}
  serviceAccount: default
  serviceAccountName: default
  terminationGracePeriodSeconds: 30
  volumes:
  - downwardAPI:
      items:
      - fieldRef:
          apiVersion: v1
          fieldPath: metadata.name
        path: podname
      - fieldRef:
          apiVersion: v1
          fieldPath: metadata.annotations
        path: annotations
    name: podinfo
```

After creating the pod, try to `kubectl edit` the pod and change the annotation value. We can find the content in the volume under `/var/lib/kubelet/pods/$uid/volumes/kubernetes.io~downward-api/podinfo/annotations` is not always changed quickly enough.

**Comment by saad-ali** [comment]
Good find @yifan-gu.

**Comment by matchstick** [comment]
@yifan-gu @saad-ali This is a great find. Is this a regression we should target into 1.3? Or safely put this in 1.4?

**Comment by matchstick** [comment]
@yifan-gu @saad-ali Marking for 1.4 just to make sure we don't lose it.

**Comment by saad-ali** [comment]
The issue resolves it self eventually (however tests will sometimes timeout before this happens). So to de-flake the test, it would be nice to get it in to a 1.3.x release, but not critical.

**Comment by Random-Liu** [comment]
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/29807/node-pull-build-e2e-test/17037/

**Comment by euank** [comment]
This also flakes a lot more in the e2es we run on CoreOS and I'm not sure why it's exacerbated so much in that configuration.

**Comment by ivan4th** [comment]
Chances are this is fixed by #29641.
At least "no such file or directory" errors look like they're caused by the race bug that PR fixes.
I've added e2e test for the race in #30674, although didn't try to reproduce it for downward API (it uses tmpfs-based emptyDir for which it's harder to induce the race).

**Comment by pwittrock** [comment]
@thockin @saad-ali @dchen1107 @pmorie @childsb 

I am leaving this in the 1.4 release as a blocking issue.  Please provide daily status updates to this issue or move it out of the release and into v1.4-nonblocking where it will no longer be tracked as a release blocking issue in the burndown meetings.

**Comment by saad-ali** [comment]
I am going to move this out of the 1.4 milestone. It is not a release blocking issue.

**Comment by davidopp** [comment]
failing in
https://k8s-testgrid.appspot.com/google-1.2-1.4-upgrade#gke-cvm-1.2-gci-1.4-upgrade-cluster

**Comment by dshulyak** [comment]
Naive fix would be to increase podLogTimeout in downstream api volume tests from 45s to, atleast as high as default reconciler loopSleepDuration, 100s. I think it will reduce frequency of flakes.

But what would be the good way to fix this? Refactor reconciler loop to work based on Watches to api-server? Marking https://github.com/kubernetes/kubernetes/issues/30189 as related

**Comment by ivan4th** [comment]
Below is the original report updated for the current master for the sake of convenience.

I believe these two flakes have the same root cause:

When there's a pod UPDATE event to kubelet, kubelet will invoke the volume manager to update the pod state to introduce the updated annotation:
https://github.com/kubernetes/kubernetes/blob/130a56ce598851e061e6bee06a5707cf00bd8621/pkg/kubelet/volumemanager/volume_manager.go#L331
The volume manager:

1, Tells the desiredStateOfWorld populator to reprocess the volume mount in the next loop.
2, Mark the volume in actualStateOfWorld to be remounted.
https://github.com/kubernetes/kubernetes/blob/130a56ce598851e061e6bee06a5707cf00bd8621/pkg/kubelet/volumemanager/volume_manager.go#L344-L345
However in the reconciler's reconciling loop, it does not always pick up the latest pod state with the latest annotation in the memory cache as the populator is running asynchronously.
https://github.com/kubernetes/kubernetes/blob/130a56ce598851e061e6bee06a5707cf00bd8621/pkg/kubelet/volumemanager/reconciler/reconciler.go#L196

But when it reads the actualStateOfWorld, it does get notified that the volume needs to be remount,
https://github.com/kubernetes/kubernetes/blob/130a56ce598851e061e6bee06a5707cf00bd8621/pkg/kubelet/volumemanager/reconciler/reconciler.go#L269

So when remounting the volume, it ends up thinking that the annotation in the pod is not changed, and thus not modifying the content on the disk.

Although the annotation updates on the disk will happen eventually after another periodic pod SYNC event received by kubelet, but it will results in the timeout of the e2e tests.

---

So basically it looks like the problem is still there unless of course I miss something.

**Comment by dims** [comment]
Can we please reopen this issue? I am seeing it regularly in my CI (Running e2e tests against local-up-cluster)

**Comment by dims** [comment]
@ivan4th : got a fix? :)

**Comment by saad-ali** [comment]
If we can get this in for 1.7 that would be great. If not, it is not a blocker for the 1.7 release, and we can move it to 1.8.

**Comment by marun** [comment]
Moving to 1.8.

**Comment by davidz627** [comment]
I recommend closing this issue as I do not believe that this race condition is causing problems anymore. The timeouts for the tests in question have been updated to have longer timeouts. And the event loop periods have been decreased to such a point that the Downward API Volume is updated in an acceptable amount of time (within 10 seconds).

We can follow the logic ivan4th has laid out and see that the race condition he refers to continues to exist. Let us note that this bad state is hit when the desired state cache is not populated with the new DownwardAPI state information before the reconciler loop is run when the DownwardAPI volume has been marked as "remount required."

Fortunately, the desired state populater runs every 100ms 
https://github.com/kubernetes/kubernetes/blob/130a56ce598851e061e6bee06a5707cf00bd8621/pkg/kubelet/volumemanager/volume_manager.go#L59
 and the reconciler loop runs every 100ms 
https://github.com/kubernetes/kubernetes/blob/130a56ce598851e061e6bee06a5707cf00bd8621/pkg/kubelet/volumemanager/volume_manager.go#L51
so we can confidently say by the second time the reconciler is run (after the potential bad state of the race condition is hit) the new DownwardAPI desired state will have been populated in the cache. However, since we already "remounted" the DownwardAPI volume in the previous loop the "remountRequired" field has been unset for the DownwardAPI and so it will not be remounted with the new desired state cache info immediately.

Fortunately, the WaitForAttachAndMount 
https://github.com/kubernetes/kubernetes/blob/130a56ce598851e061e6bee06a5707cf00bd8621/pkg/kubelet/volumemanager/volume_manager.go#L331
method is called on a timer every 3-10 seconds (determined empirically, could not find exact timer logic). As soon as this method is called, the DownwardAPI is marked again for remount, the new info would have been populated into the desired state cache, and the DownwardAPI is successfully remounted with the updated objects.

Therefore we can expect any of the UPDATE calls to kublet to be reflected in the mounted DownwardAPI volume in a maximum of around 10 seconds. We have determined this to be an acceptable amount of time for the update to be reflected in the volume.

**Comment by davidz627** [comment]
I recommend closing this issue as I do not believe that this race condition is causing problems anymore. The timeouts for the tests in question have been updated to have longer timeouts. And the event loop periods have been decreased to such a point that the Downward API Volume is updated in an acceptable amount of time (within 1 minute).

We can follow the logic ivan4th has laid out and see that the race condition he refers to continues to exist. Let us note that this bad state is hit when the desired state cache is not populated with the new DownwardAPI state information before the reconciler loop is run when the DownwardAPI volume has been marked as "remount required."

Fortunately, the desired state populater runs every 100ms 
https://github.com/kubernetes/kubernetes/blob/130a56ce598851e061e6bee06a5707cf00bd8621/pkg/kubelet/volumemanager/volume_manager.go#L59
 and the reconciler loop runs every 100ms 
https://github.com/kubernetes/kubernetes/blob/130a56ce598851e061e6bee06a5707cf00bd8621/pkg/kubelet/volumemanager/volume_manager.go#L51
so we can confidently say by the second time the reconciler is run (after the potential bad state of the race condition is hit) the new DownwardAPI desired state will have been populated in the cache. However, since we already "remounted" the DownwardAPI volume in the previous loop the "remountRequired" field has been unset for the DownwardAPI and so it will not be remounted with the new desired state cache info immediately.

Fortunately, the WaitForAttachAndMount 
https://github.com/kubernetes/kubernetes/blob/130a56ce598851e061e6bee06a5707cf00bd8621/pkg/kubelet/volumemanager/volume_manager.go#L331
method is called for each pod on a timer every minute
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/kubeletconfig/v1alpha1/defaults.go#L216
As soon as this method is called, the DownwardAPI is marked again for remount, the new info would have been populated into the desired state cache, and the DownwardAPI is successfully remounted with the updated objects.

Therefore we can expect any of the UPDATE calls to kublet to be reflected in the mounted DownwardAPI volume in a maximum of around 1 minute. We have determined this to be an acceptable amount of time for the update to be reflected in the volume.
